src/server/game/AI/CreatureAI.cpp

```diff
@@ -271,6 +271,112 @@ bool CreatureAI::_EnterEvadeMode()
     return true;
 }
 
+#define BOUNDARY_VISUALIZE_CREATURE 15425
+#define BOUNDARY_VISUALIZE_CREATURE_SCALE 0.25f
+#define BOUNDARY_VISUALIZE_STEP_SIZE 1
+#define BOUNDARY_VISUALIZE_FAILSAFE_LIMIT 200
+#define BOUNDARY_VISUALIZE_SPAWN_HEIGHT 5
+void CreatureAI::VisualizeBoundary(uint32 duration, Unit* owner, bool fill) const
+{
+    typedef std::pair<int32, int32> coordinate;
+    
+    if (!owner)
+        return;
+
+    if (!_boundary || _boundary->empty())
+    {
+        me->Talk("VisualizeBoundary error: Creature movement is unbounded", CHAT_MSG_MONSTER_YELL, LANG_UNIVERSAL, 50000.0f, nullptr);
+        return;
+    }
+
+    std::queue<coordinate> Q;
+    std::unordered_set<coordinate> alreadyChecked;
+    std::unordered_set<coordinate> outOfBounds;
+
+    float meX = me->GetPositionX(), meY = me->GetPositionY(), meZ = me->GetPositionZ(), spawnZ = meZ + BOUNDARY_VISUALIZE_SPAWN_HEIGHT;
+    
+    if (!CheckBoundary(me))
+    {
+        me->Talk("VisualizeBoundary warning: Current position is out of bounds", CHAT_MSG_MONSTER_YELL, LANG_UNIVERSAL, 50000.0f, nullptr);
+        return;
+    }
+    bool hadBoundsWarning = false;
+    Q.push({ 0,0 });
+    while (!Q.empty())
+    {
+        coordinate front = Q.front();
+        bool hasOutOfBoundsNeighbor = false;
+        for (coordinate off : std::list<coordinate>({ {1, 0}, { 0,1 }, { -1,0 }, { 0,-1 } }))
+        {
+            coordinate next(front.first + off.first, front.second + off.second);
+            if (next.first > BOUNDARY_VISUALIZE_FAILSAFE_LIMIT || next.first < -BOUNDARY_VISUALIZE_FAILSAFE_LIMIT || next.second > BOUNDARY_VISUALIZE_FAILSAFE_LIMIT || next.second < -BOUNDARY_VISUALIZE_FAILSAFE_LIMIT)
+            {
+                if (!hadBoundsWarning)
+                {
+                    hadBoundsWarning = true;
+                    me->Talk("VisualizeBoundary warning: Reached fail-safe flood boundary - check is your boundary is unbounded!", CHAT_MSG_MONSTER_YELL, LANG_UNIVERSAL, 50000.0f, nullptr);
+                }
+                continue;
+            }
+            if (alreadyChecked.find(next) == alreadyChecked.end()) // never check a coordinate twice
+            {
+                Position nextPos(meX + next.first*BOUNDARY_VISUALIZE_STEP_SIZE, meY + next.second*BOUNDARY_VISUALIZE_STEP_SIZE, meZ);
```

Comment:
Would need to turn the Vector3D back to a Position though for boundary check. Probably not worth it overhead wise?